crates/sui-storage/src/write_ahead_log.rs

```diff
@@ -0,0 +1,291 @@
+    #[must_use]
+    async fn begin_tx(
+        &'a self,
+        tx: &TransactionDigest,
+        cert: &C,
+    ) -> SuiResult<Option<DBTxGuard<'a, C>>> {
+        let _mutex_guard = self.mutex_table.acquire_lock(tx).await;
+
+        if self.log.contains_key(tx)? {
+            // We return None instead of a guard, to signal that a tx with this digest is already
+            // in progress.
+            //
+            // TODO: It may turn out to be better to hold the lock until the other guard is
+            // dropped - this should become clear once this code is being used.
+            return Ok(None);
+        }
+
+        self.log.insert(tx, cert)?;
+
+        Ok(Some(DBTxGuard::new(tx, self)))
+    }
+
+    #[must_use]
+    fn take_recoverable_txes(&'a self) -> Vec<DBTxGuard<'a, C>> {
+        // unwrap ok because we should absolutely crash if the mutex is poisoned
+        let mut v = self.recoverable_txes.lock().unwrap();
+        let mut new = Vec::new();
+        std::mem::swap(&mut *v, &mut new);
+        let ret = new;
+        ret.iter()
+            .map(|digest| DBTxGuard::new(digest, self))
+            .collect()
+    }
+
+    fn get_tx_data(&self, g: &DBTxGuard<'a, C>) -> SuiResult<C> {
+        self.log
+            .get(&g.tx)
+            .map_err(SuiError::from)?
+            .ok_or(SuiError::TransactionNotFound { digest: g.tx })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+
+    use crate::write_ahead_log::{DBWriteAheadLog, TxGuard, WriteAheadLog};
+    use anyhow;
+    use sui_types::base_types::TransactionDigest;
+
+    #[tokio::test]
+    async fn test_write_ahead_log() -> Result<(), anyhow::Error> {
+        let working_dir = tempfile::tempdir()?;
+
+        let tx1_id = TransactionDigest::random();
+        let tx2_id = TransactionDigest::random();
+        let tx3_id = TransactionDigest::random();
+
+        {
+            let log: DBWriteAheadLog<u32> = DBWriteAheadLog::new(&working_dir);
+            let r = log.take_recoverable_txes();
+            assert!(r.is_empty());
+
+            let tx1 = log.begin_tx(&tx1_id, &1).await?.unwrap();
+            tx1.commit_tx().unwrap();
+
+            let tx2 = log.begin_tx(&tx2_id, &2).await?.unwrap();
+            tx2.commit_tx().unwrap();
+
+            {
+                let _tx3 = log.begin_tx(&tx3_id, &3).await?.unwrap();
+                // implicit drop
+            }
+
+            let r = log.take_recoverable_txes();
+            // tx3 in recoverable txes because we dropped the guard.
+            assert_eq!(r.len(), 1);
+            assert_eq!(r[0].tx_id(), tx3_id);
+
+            // verify previous call emptied the recoverable list
+            let r_empty = log.take_recoverable_txes();
+            assert!(r_empty.is_empty());
+        }
+
+        {
+            // recover the log
+            let log: DBWriteAheadLog<u32> = DBWriteAheadLog::new(&working_dir);
+
+            // recoverable txes still there
+            let mut r = log.take_recoverable_txes();
```

Comment:
I thought the previous call to `take_recoverable_txes` clears the table?  Wouldn't have that been committed to disk already?